src/Compilers/Core/Portable/DiagnosticAnalyzer/IAnalyzerAssemblyResolver.cs

```diff
@@ -2,21 +2,39 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#if NET
+
 using System.Reflection;
+using System.Runtime.Loader;
 
 namespace Microsoft.CodeAnalysis
 {
     /// <summary>
-    /// Allows a host to override how assembly resolution is performed by the <see cref="AnalyzerAssemblyLoader"/>.
+    /// This interface allows hosts to control exactly how a given <see cref="AssemblyName"/> is resolved to an
+    /// <see cref="Assembly"/> instance. This is useful for hosts that need to load assemblies in a custom way like
+    /// Razor or stream based loading.
     /// </summary>
     internal interface IAnalyzerAssemblyResolver
     {
         /// <summary>
-        /// Attempts to resolve an assembly by name.
+        /// Resolve an <see cref="Assembly"/> for the given parameters.
         /// </summary>
-        /// <param name="assemblyName">The assembly to resolve.</param>
-        /// <param name="rootDirectory">The root directory the calling <see cref="AnalyzerAssemblyLoader"/> is configured with.</param>
-        /// <returns>The resolved assembly, or <see langword="null"/></returns>
-        Assembly? ResolveAssembly(AssemblyName assemblyName, string rootDirectory);
+        /// <remarks>
+        /// The <see cref="AnalyzerAssemblyLoader"/> will partition analyzers into the directories they live
+        /// in and will create a separate <see cref="AssemblyLoadContext"/> for each directory. That instance
+        /// and the directory name represent <paramref name="directoryContext" /> and <paramref name="directory" />.
+        /// 
+        /// This is invoked as part of <see cref="AssemblyLoadContext.Load(AssemblyName)"/>. Exceptions in
+        /// the implementation of this interface will escape from that method and be registered as the result
+        /// of load.
+        /// </remarks>
+        /// <param name="loader">The <see cref="AnalyzerAssemblyLoader"/> instance that is performing the load</param>
+        /// <param name="assemblyName">The name of the assembly to be loaded</param>
+        /// <param name="directoryContext">The <see cref="AssemblyLoadContext"/> for the <paramref name="directory"/></param>
+        /// <param name="directory">The directory where the assembly is being loaded from</param>
```

Comment:
Might be good to clarify if this is the "original" directory or one that has gone through mapping/shadow copying, since I suspect that's why you're doing directory vs. directoryContext here.